File: rust_exercises/leetcode/Problem716/MaxStack.java
class MaxStack {
    Stack<Integer> stack;
    Stack<Integer> maxStack;

    /** initialize your data structure here. */
    public MaxStack() {
        stack = new Stack();
        maxStack = new Stack();
    }
    
    public void push(int x) {
        stack.push(x);
        if (maxStack.isEmpty()) {
            maxStack.push(x);
        } else {
            maxStack.push(Math.max(maxStack.peek(), x));
        }
    }
    
    public int pop() {
        maxStack.pop();
        return stack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int peekMax() {
        return maxStack.peek();
    }
    
    public int popMax() {
        Stack<Integer> temp = new Stack<>();
        int max = maxStack.peek();
        while (stack.peek() != max) {
            temp.push(pop());
        }
        pop();
        while (!temp.isEmpty()) {
            push(temp.pop());
        }
        return max;
    }
}

/**
 * Your MaxStack object will be instantiated and called as such:
 * MaxStack obj = new MaxStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.peekMax();
 * int param_5 = obj.popMax();
 */

File: rust_exercises/leetcode/Problem716/max_stack.rs
struct MaxStack {
    stack: Vec<i32>,
    max_stack: Vec<i32>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MaxStack {

    /** initialize your data structure here. */
    fn new() -> Self {
        MaxStack {
            stack: vec![],
            max_stack: vec![],
        }
    }
    
    fn push(&mut self, x: i32) {
        self.stack.push(x);
        if self.max_stack.len() == 0 {
            self.max_stack.push(x);
        } else {
            let max = self.max_stack[self.max_stack.len() - 1];
            self.max_stack.push(std::cmp::max(x, max));
        }
    }
    
    fn pop(&mut self) -> i32 {
        self.max_stack.pop();
        self.stack.pop().unwrap()
    }
    
    fn top(&self) -> i32 {
        self.stack[self.stack.len() - 1]
    }
    
    fn peek_max(&self) -> i32 {
        self.max_stack[self.stack.len() - 1]
    }
    
    fn pop_max(&mut self) -> i32 {
        let max = self.max_stack[self.stack.len() - 1];
        let mut temp: Vec<i32> = vec![];
        loop {
            let x = self.pop();
            if x == max {
                break;
            }
            temp.push(x);
        }
        while temp.len() > 0 {
            self.push(temp.pop().unwrap());
        }
        max
    }
}

/**
 * Your MaxStack object will be instantiated and called as such:
 * let obj = MaxStack::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.peek_max();
 * let ret_5: i32 = obj.pop_max();
 */
